frontend/include/chpl/framework/error-classes-list.h

```diff
@@ -26,100 +26,111 @@
 //
 // Calls to the ERROR_CLASS/WARNING_CLASS/etc. macros should list the new
 // error's name first, followed a list of types describing error details.
+//
+// Errors are grouped by which compilation stage they relate to, and sorted
+// alphabetically within these groups. Groups themselves are ordered by the
+// order their compilation stages occur in.
 
-ERROR_CLASS(IncompatibleIfBranches,
-    const uast::Conditional*,
-    types::QualifiedType,
-    types::QualifiedType)
-ERROR_CLASS(TupleExpansionNamedArgs, const uast::OpCall*, const uast::FnCall*)
-ERROR_CLASS(MemManagementNonClass, const uast::New*, const types::Type*)
-ERROR_CLASS(PrivateToPublicInclude, const uast::Include*, const uast::Module*)
-ERROR_CLASS(PrototypeInclude, const uast::Include*, const uast::Module*)
-ERROR_CLASS(MissingInclude, const uast::Include*, std::string)
-ERROR_CLASS(UseImportUnknownSym, const uast::VisibilityClause*,
-            const resolution::Scope*, const resolution::VisibilityStmtKind,
-            std::string)
-ERROR_CLASS(UseImportUnknownMod, const ID, const resolution::VisibilityStmtKind,
-            std::string)
-ERROR_CLASS(UseImportNotModule, const ID, const resolution::VisibilityStmtKind,
-            std::string)
-ERROR_CLASS(AsWithUseExcept, const uast::Use*, const uast::As*)
-ERROR_CLASS(DotExprInUseImport, const uast::VisibilityClause*,
-            const uast::VisibilityClause::LimitationKind, const uast::Dot*)
-ERROR_CLASS(UnsupportedAsIdent, const uast::As*, const uast::AstNode*)
-ERROR_CLASS(Redefinition, const uast::NamedDecl*, std::vector<ID>)
+/* begin parser errors */
+PARSER_SYNTAX_CLASS(CannotAttachPragmas, const uast::AstNode*)
+PARSER_SYNTAX_CLASS(ExceptOnlyInvalidExpr,
+                    uast::VisibilityClause::LimitationKind)
+PARSER_SYNTAX_CLASS(InvalidIndexExpr)
+PARSER_SYNTAX_CLASS(InvalidNewForm, const uast::AstNode*)
+PARSER_SYNTAX_CLASS(InvalidNumericLiteral, std::string)
+PARSER_SYNTAX_CLASS(LabelIneligibleStmt, const uast::AstNode*)
+PARSER_ERROR_CLASS(MultipleExternalRenaming)
+PARSER_SYNTAX_CLASS(NewWithoutArgs, const uast::AstNode*)
+// ParseErr and ParseSyntax are catch-alls for simple parsing errors that do not
+// have a specialized error class
+PARSER_ERROR_CLASS(ParseErr, std::string)
+PARSER_SYNTAX_CLASS(ParseSyntax, std::string)
+PARSER_WARNING_CLASS(PreIncDecOp, bool)
+PARSER_ERROR_CLASS(RecordInheritanceNotSupported, std::string)
+PARSER_SYNTAX_CLASS(UseImportNeedsModule, bool)
+/* begin Bison-specific parser errors */
```

Comment:
Since the names are already prefixed with the word "Bison", I'd actually merge them with the usual "parser errors" category and remove the "begin/end parser specific errors" comments.